src/model/openthaigpt_pretraining_model/lightning/utils.py

```diff
@@ -0,0 +1,219 @@
+                LANGUAGE_DATASET,
+                streaming=True,
+                split=SPLIT_VAL,
+            )
+        elif mode == "train":
+            self.data_set = load_dataset(
+                DATASET_NAME,
+                LANGUAGE_DATASET,
+                streaming=True,
+                split=SPLIT_TRAIN,
+            ).shuffle(buffer_size=10_000)
+        else:
+            raise NotImplementedError("only support Train,Val")
+
+    def __iter__(self):
+        buffer = []
+        iter_dataset = self.data_set
+
+        for sample in iter_dataset:
+            buffer += self.tokenizer(sample["text"])["input_ids"]
+            while len(buffer) > self.max_tokens:
+                yield torch.tensor(buffer[: self.max_tokens])
+                buffer = buffer[self.max_tokens :]
+
+
+def seed_everything(seed):
+    torch.manual_seed(seed)
+    torch.cuda.manual_seed(seed)
+    torch.cuda.manual_seed_all(seed)
+    torch.backends.cudnn.deterministic = True
+    torch.backends.cudnn.benchmark = False
+    np.random.seed(seed)
+    random.seed(seed)
+
+
+class Trainer:
+    def __init__(
+        self,
+        accelerator: Union[str, Accelerator] = "auto",
+        strategy: Union[str, Strategy] = "auto",
+        devices: Union[List[int], str, int] = "auto",
+        precision: Union[str, int] = "32-true",
+        seed: int = 42,
+        batch_size: int = 8,
+        grad: int = 4,
+        context_length: int = 256,
+        model_name: str = "llama",
+        optimizer: str = "adamw",
+        weight_decay: float = 1e-2,
+        lr: float = 1e-4,
+    ):
+        self.max_tokens = context_length
+        self.step = 0
+        self.seed = seed
+        self.grad = grad
+        self.fabric = L.Fabric(
+            accelerator=accelerator,
+            strategy=strategy,
+            devices=devices,
+            precision=precision,
+        )
+        self.fabric.launch()
+        if model_name == "llama":
+            model_name = LLAMA_MODEL  # for tokenizer
+            # cfg = ModelArgs(
+            #     dim=512,
+            #     n_layers=8,
+            #     n_heads=8,
+            #     vocab_size=32000,
+            #     multiple_of=256,
+            #     norm_eps=1e-5,
+            #     max_batch_size=32,
+            #     max_seq_len=2048,
+            #     attention_mode=ORIGIN_ATTENTION_MODE,
+            # )
+            # self.model = model = Transformer(cfg)
+            cfg = LlamaConfig(
+                vocab_size=32000,
+                hidden_size=1024,
+                num_hidden_layers=8,
+                num_attention_heads=8,
+                hidden_act="silu",
+                max_position_embeddings=2048,  # sequence length
+            )
+            self.model = model = LlamaForCausalLM(cfg)
+        elif model_name == "gptj":
+            model_name = GPTJ_MODEL  # for tokenizer
+            cfg = GPTJConfig(
+                vocab_size=50400,
+                n_positions=2048,
```

Comment:
n_positions should be equal to context length

Reply:
done